File: articles/problem-framing-design-thinking/cpp/problem_frame_value_engine.cpp
#include <algorithm>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

struct ProblemFrame {
    std::string name;
    double explanatory_adequacy;
    double stakeholder_coverage;
    double opportunity_value;
    double framing_risk;
};

struct ScoredFrame {
    std::string name;
    double value;
};

double to_double(const std::string& value) {
    try {
        return std::stod(value);
    } catch (...) {
        throw std::runtime_error("Unable to parse numeric value: " + value);
    }
}

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> fields;
    std::stringstream ss(line);
    std::string item;

    while (std::getline(ss, item, ',')) {
        fields.push_back(item);
    }

    return fields;
}

std::vector<ProblemFrame> read_frames(const std::string& path) {
    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("Unable to open input file: " + path);
    }

    std::vector<ProblemFrame> frames;
    std::string line;

    std::getline(file, line); // header

    while (std::getline(file, line)) {
        if (line.empty()) {
            continue;
        }

        auto fields = split_csv_line(line);
        if (fields.size() < 5) {
            throw std::runtime_error("Malformed CSV row: " + line);
        }

        frames.push_back(
            ProblemFrame{
                fields[0],
                to_double(fields[1]),
                to_double(fields[2]),
                to_double(fields[3]),
                to_double(fields[4])
            }
        );
    }

    return frames;
}

double frame_value(const ProblemFrame& frame) {
    constexpr double we = 0.30;
    constexpr double ws = 0.25;
    constexpr double wo = 0.30;
    constexpr double wr = 0.15;

    return we * frame.explanatory_adequacy
         + ws * frame.stakeholder_coverage
         + wo * frame.opportunity_value
         - wr * frame.framing_risk;
}

int main(int argc, char* argv[]) {
    const std::string default_path = "../data/raw/problem_frames_raw.csv";
    const std::string input_path = argc > 1 ? argv[1] : default_path;

    try {
        auto frames = read_frames(input_path);
        std::vector<ScoredFrame> scored;

        for (const auto& frame : frames) {
            scored.push_back(ScoredFrame{frame.name, frame_value(frame)});
        }

        std::sort(
            scored.begin(),
            scored.end(),
            [](const ScoredFrame& a, const ScoredFrame& b) {
                return a.value > b.value;
            }
        );

        std::cout << "rank,frame,frame_value\n";
        for (std::size_t i = 0; i < scored.size(); ++i) {
            std::cout << (i + 1) << ","
                      << scored[i].name << ","
                      << std::fixed << std::setprecision(6)
                      << scored[i].value << "\n";
        }
    } catch (const std::exception& ex) {
        std::cerr << "ERROR: " << ex.what() << "\n";
        return 1;
    }

    return 0;
}

File: articles/problem-framing-design-thinking/go/problem_frame_value_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type ProblemFrame struct {
	Name                 string
	ExplanatoryAdequacy  float64
	StakeholderCoverage  float64
	OpportunityValue     float64
	FramingRisk          float64
	Value                float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func frameValue(frame ProblemFrame) float64 {
	return 0.30*frame.ExplanatoryAdequacy +
		0.25*frame.StakeholderCoverage +
		0.30*frame.OpportunityValue -
		0.15*frame.FramingRisk
}

func loadFrames(path string) []ProblemFrame {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	var frames []ProblemFrame

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 5 {
			log.Fatalf("malformed row: %v", row)
		}

		frame := ProblemFrame{
			Name:                row[0],
			ExplanatoryAdequacy: parseFloat(row[1]),
			StakeholderCoverage: parseFloat(row[2]),
			OpportunityValue:    parseFloat(row[3]),
			FramingRisk:         parseFloat(row[4]),
		}
		frame.Value = frameValue(frame)
		frames = append(frames, frame)
	}

	return frames
}

func main() {
	path := "../data/raw/problem_frames_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	frames := loadFrames(path)

	sort.Slice(frames, func(i, j int) bool {
		return frames[i].Value > frames[j].Value
	})

	fmt.Println("rank,frame,frame_value")
	for i, frame := range frames {
		fmt.Printf("%d,%s,%.6f\n", i+1, frame.Name, frame.Value)
	}
}
